operator/scheduler/model_test.go

```diff
@@ -0,0 +1,256 @@
+								Namespace:  "default",
+								Generation: 1,
+							},
+							State: &scheduler.ModelStatus{
+								State:             scheduler.ModelStatus_ModelTerminating,
+								AvailableReplicas: 1,
+							},
+						},
+					},
+				},
+			},
+		},
+	}
+
+	for _, test := range tests {
+		t.Run(test.name, func(t *testing.T) {
+			// note that if responses_pipelines is nil -> scheduler state is not existing
+			var grpcClient mockSchedulerGrpcClient
+			if !test.noSchedulerState {
+				grpcClient = mockSchedulerGrpcClient{
+					responses_subscribe_models: test.results,
+					responses_models:           test.results,
+				}
+			} else {
+				grpcClient = mockSchedulerGrpcClient{
+					responses_subscribe_models: test.results,
+				}
+			}
+			controller := newMockControllerClient(test.existing_resources...)
+			err := controller.handleModels(context.Background(), &grpcClient, "")
+			g.Expect(err).To(BeNil())
+			err = controller.SubscribeModelEvents(context.Background(), &grpcClient, "")
+			g.Expect(err).To(BeNil())
+
+			isBeingDeleted := map[string]bool{}
+			for _, req := range test.existing_resources {
+				if !req.GetDeletionTimestamp().IsZero() {
+					isBeingDeleted[req.GetName()] = true
+				} else {
+					isBeingDeleted[req.GetName()] = false
+				}
+			}
+
+			// for model resources that are not deleted we reload them
+			// this is not necessary check but added for sanity
+			activeResources := 0
+			for idx, req := range test.existing_resources {
+				if req.GetDeletionTimestamp().IsZero() {
+					g.Expect(req.GetName()).To(Equal(grpcClient.requests_models[idx].Model.GetMeta().GetName()))
+					activeResources++
+				}
+			}
+			g.Expect(len(grpcClient.requests_models)).To(Equal(activeResources))
+
+			// check state is correct for each model
+			for _, r := range test.results {
+				if r.Versions[0].State.GetState() != scheduler.ModelStatus_ModelTerminated {
+					model := &mlopsv1alpha1.Model{}
+					err := controller.Get(
+						context.Background(),
+						client.ObjectKey{
+							Name:      r.GetModelName(),
+							Namespace: r.Versions[0].KubernetesMeta.Namespace,
+						},
+						model,
+					)
+					// we check if the model is not in k8s (existing_resources) then we should not act on it
+					if _, ok := isBeingDeleted[r.GetModelName()]; !ok {
+						g.Expect(err).ToNot(BeNil())
+					} else {
+						g.Expect(err).To(BeNil())
+					}
+					if r.Versions[0].State.GetState() == scheduler.ModelStatus_ModelAvailable {
+						g.Expect(model.Status.IsReady()).To(BeTrueBecause("Model state is ModelAvailable"))
+					} else {
+						g.Expect(model.Status.IsReady()).To(BeFalseBecause("Model state is not ModelAvailable"))
+					}
+
+					g.Expect(uint32(model.Status.Replicas)).To(Equal(r.Versions[0].State.GetAvailableReplicas() + r.Versions[0].State.GetUnavailableReplicas()))
+					g.Expect(model.Status.Selector).To(Equal("server=" + r.Versions[0].ServerName))
+				} else {
+					model := &mlopsv1alpha1.Pipeline{}
```

Comment:
Shouldn't this be `&mlopsv1alpha1.Model{}`? I think this shows that the simple expect on `err != nil` is perhaps less strong than needed for some cases.

Reply:
Good catch, fixed.